### PR TITLE
Wip new

### PR DESCRIPTION
**Pull request by sullysaurus**
@jmburges @AnnJohn @SophieDeBenedetto I need some readme guidance since this is a review, should it look like this https://github.com/learn-co-curriculum/school-domain-with-db/blob/solution/README.md one?

### CODE REVIEW

File: spec/dog_spec.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  describe "#drop_table" do
+    it 'drops the dogs table from the database' do
+      Dog.drop_table
+      table_check_sql = "SELECT tbl_name FROM sqlite_master WHERE type='table' AND tbl_name='dogs';"
+      expect(DB[:conn].execute(table_check_sql)[0]).to eq(nil)
+    end
+  end
+
+  describe "#save" do
```
Comment (jmburges) [inline spec/dog_spec.rb:48]: Also include a test that will do an update if the object has already been persisted

File: spec/dog_spec.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    it 'saves an instance of the dog class to the database and then sets the given dogs `id` attribute' do
+      teddy.save
+      expect(DB[:conn].execute("SELECT * FROM dogs")).to eq([[1, "Teddy", "cockapoo"]])
+      expect(teddy.id).to eq(1)
+    end
+  end
+
+  describe "#create" do
+    it 'takes in a hash of attributes and uses metaprogramming to create a new dog object. Then it uses the #save method to save that dog to the database' do
+      Dog.create(name: "Ralph", breed: "lab")
```
Comment (jmburges) [inline spec/dog_spec.rb:58]: also create should return a new fully formed `Dog` object. Check that. Probably in a different `it` block

File: spec/dog_spec.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      expect(DB[:conn].execute("SELECT * FROM dogs")).to eq([[1, "Ralph", "lab"]])
+    end
+  end
+
+  describe '.find_or_create_by' do
+    it 'creates an instance of a dog if it does not already exist' do
+      dog1 = Dog.create(name: 'teddy', breed: 'cockapoo')
+
+      dog2 = Dog.find_or_create_by(name: 'teddy', breed: 'cockapoo')
+      expect(dog1.id).to eq(dog2.id)
```
Comment (jmburges) [inline spec/dog_spec.rb:84]: handful more tests needed here:
- [x] Creates one if it doesn't exist
- [x] Add in two dogs with the same name, different breed, then do a `find_or_create_by` on the name and one of the breeds, make sure the id is correct and it returns the right one!
- [x] Add in two dogs with the same name, different breed, then `find_or_create_by` another dog with the same name and a different breed, make sure it's a new one

Comment (jmburges) [inline spec/dog_spec.rb:84]: we should add that they need to write a `find_by_id` method

File: spec/dog_spec.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      row = [1, "Pat", "poodle"]
+      pat = Dog.new_from_db(row)
+
+      expect(pat.id).to eq(row[0])
+      expect(pat.name).to eq(row[1])
+      expect(pat.breed).to eq(row[2])
+    end
+  end
+
+  describe '#find_by_name' do
```
Comment (jmburges) [inline spec/dog_spec.rb:116]: descrbe `.find_by_name` it's a class method

File: spec/dog_spec.rb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    it 'returns an instance of student that matches the name from the DB' do
+      teddy.save
+      teddy_from_db = Dog.find_by_name("Teddy")
+      expect(teddy_from_db.name).to eq("Teddy")
```
Comment (jmburges) [inline spec/dog_spec.rb:121]: check ids as well to be sure it's the same row

**Comment by jmburges** [comment]
Made a bunch of comments. It also needs a README.md. It should be similar to the school domain one but no need to explain what ORMs are at all. Just explain the class they'll be making and tell them to read the specs. Give them a bit of guidance on which methods to attack first

File: spec/dog_spec.rb (hunk 2 above)
Comment (sullysaurus) [inline spec/dog_spec.rb:58]: Updated create method to return object.

File: spec/dog_spec.rb (hunk 1 above)
Comment (sullysaurus) [inline spec/dog_spec.rb:48]: Updated.

File: spec/dog_spec.rb (hunk 2 above)
Comment (sullysaurus) [inline spec/dog_spec.rb:58]: Added test for it.

File: spec/dog_spec.rb (hunk 4 above)
Comment (sullysaurus) [inline spec/dog_spec.rb:116]: Updated.

File: spec/dog_spec.rb (hunk 5 above)
Comment (sullysaurus) [inline spec/dog_spec.rb:121]: Updated.

**Comment by jmburges** [comment]
great!

**Pull request merged**